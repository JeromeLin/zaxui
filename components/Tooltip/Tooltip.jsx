import React, { PureComponent, cloneElement } from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import classnames from 'classnames';

class Tooltip extends PureComponent {

  constructor(props) {
    super(props);
    this.state = {

    };
  }

  show() {
    const { prefixCls, className, message } = this.props;

    const cls = classnames({
      [`${prefixCls}`]: true,
      [className]: !!className,
    });

    const child = this.child.getBoundingClientRect();
    const style = {
      left: child.left,
      top: child.top,
    };

    const root = document.createElement('div');
    root.className = `${prefixCls}-wrapper`;
    document.body.appendChild(root);
    ReactDOM.render(<div className={cls} style={style}>{message}</div>, root);
  }

  render() {
    const { children } = this.props;

    return cloneElement(children, {
      ref: ele => this.child = ele,
    });
  }
}

Tooltip.propTypes = {
  prefixCls: PropTypes.string,
  className: PropTypes.string,
};

Tooltip.defaultProps = {
  prefixCls: 'za-tooltip',
};

export default Tooltip;
