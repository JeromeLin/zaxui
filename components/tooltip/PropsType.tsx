import { ReactNode } from 'react';

export type direction =
  | 'topLeft'
  | 'top'
  | 'topRight'
  | 'rightTop'
  | 'right'
  | 'rightBottom'
  | 'bottomLeft'
  | 'bottom'
  | 'bottomRight'
  | 'leftTop'
  | 'left'
  | 'leftBottom';

export default interface PropsType {
  prefixCls?: string;
  className?: string;
  trigger?: 'click' | 'hover';
  visible?: boolean;
  direction: direction;
  onVisibleChange?: (visible?: boolean) => void;
  title?: () => void | ReactNode ;
}
