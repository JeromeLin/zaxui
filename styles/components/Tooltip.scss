$prefixCls: za-tooltip;

.#{$prefixCls} {
  position: absolute;
  top: r(20);
  min-height: r(21);
  padding: r(5);
  background-color: rgba(0, 0, 0, .7);
  border-radius: $radius-md;
  color: #fff;
  font-size: r(12);

  &-wrapper {
    position: absolute;
    left: 0;
    top: 0;
    width: 100%;
  }
}