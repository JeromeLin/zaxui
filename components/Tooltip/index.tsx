import React, { PureComponent, cloneElement } from 'react';
import ReactDOM, { findDOMNode } from 'react-dom';
import classnames from 'classnames';
import { TooltipProps } from './PropsType';
// import Events from '../utils/events';

export default class Tooltip extends PureComponent<TooltipProps, any> {

  private child;
  private tooltip;

  static defaultProps = {
    prefixCls: 'za-tooltip',
    visible: false,
  }

  componentDidMount() {
    if (!window.zarmTooltip) {
      window.zarmTooltip = document.createElement('div');
      document.body.appendChild(window.zarmTooltip);
    }
    this.show(this.props);
    // Events.on(window, 'resize', this.resize);
  }

  componentWillReceiveProps(nextProps) {
    this.show(nextProps);
  }

  show = (props) => {
    const { prefixCls, className, visible, message } = props;
    const cls = classnames(prefixCls, className, {
      [`${prefixCls}-hidden`]: !visible,
    });

    // eslint-disable-next-line
    const rect = findDOMNode(this.child).getBoundingClientRect();
    const scrollTop = document.documentElement.scrollTop + document.body.scrollTop;
    const style = {
      left: rect.left,
      top: rect.top + scrollTop,
      width: rect.width,
    };

    ReactDOM.render(
      <div className={cls} style={style} ref={(ele) => { this.tooltip = ele; }}>
        <div className={`${prefixCls}-inner`}>{message}</div>
      </div>
    , window.zarmTooltip);
  }

  render() {
    const { children } = this.props;

    return cloneElement(children, {
      ref: (ele) => { this.child = ele; },
    });
  }
}

