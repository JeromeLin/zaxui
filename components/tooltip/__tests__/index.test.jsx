import React from 'react';
import { render, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Tooltip from '../index';

describe('Tooltip', () => {
  it('visible', () => {
    jest.useFakeTimers();
    const wrapper = mount(<Tooltip visible title="hello"><span>foo</span></Tooltip>);
    expect(toJson(wrapper)).toMatchSnapshot();
    jest.runAllTimers();
    wrapper.unmount();
  });

  it('direction', () => {
    jest.useFakeTimers();
    const wrapper = mount(<Tooltip direction="top" title="hello"><span>foo</span></Tooltip>);
    expect(toJson(wrapper)).toMatchSnapshot();
    jest.runAllTimers();
    wrapper.unmount();
  });
});
