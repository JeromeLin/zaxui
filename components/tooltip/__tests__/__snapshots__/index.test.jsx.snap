// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip direction 1`] = `
<Tootip
  direction="top"
  onVisibleChange={[Function]}
  prefixCls="za-tooltip"
  title="hello"
  trigger="hover"
  visible={false}
>
  <div
    className="za-tooltip"
  >
    <Portal
      containerInfo={
        <body>
          <div
            class="za-tooltip__content"
          >
            hello
            <span
              class="za-tooltip__arrow"
              id="za-tooltip-arrow"
            />
          </div>
        </body>
      }
    >
      <div
        className="za-tooltip__content"
      >
        hello
        <span
          className="za-tooltip__arrow"
          id="za-tooltip-arrow"
        />
      </div>
    </Portal>
    <span
      className="za-tooltip__inner"
    >
      <span>
        foo
      </span>
    </span>
  </div>
</Tootip>
`;

exports[`Tooltip visible 1`] = `
<Tootip
  direction="top"
  onVisibleChange={[Function]}
  prefixCls="za-tooltip"
  title="hello"
  trigger="hover"
  visible={true}
>
  <div
    className="za-tooltip"
  >
    <Portal
      containerInfo={
        <body>
          <div
            class="za-tooltip__content za-tooltip__content--show"
            style="position: fixed; top: 0px; transform: translate3d(NaNpx, NaNpx, 0); left: 0px;"
            x-placement="top"
          >
            hello
            <span
              class="za-tooltip__arrow"
              id="za-tooltip-arrow"
              style=""
            />
          </div>
        </body>
      }
    >
      <div
        className="za-tooltip__content za-tooltip__content--show"
      >
        hello
        <span
          className="za-tooltip__arrow"
          id="za-tooltip-arrow"
        />
      </div>
    </Portal>
    <span
      className="za-tooltip__inner"
    >
      <span>
        foo
      </span>
    </span>
  </div>
</Tootip>
`;
